### DISCUSSION TITLE
New API for Nexus Portal Integration with Icon Bridge

### DISCUSSION THREAD

**Discussion by bbist** [icon-project/icon-bridge#68] [General]
Because of a need to make the service handlers consistent, we decided to merge `nativecoin bsh` and `token bsh`. Now, we have a single BTPTokenService (BTS) that serves the same purpose of aforementioned two contracts. Here are the transfer cases:

1. ICON -> BSC:
  - NativeCoin (ICX):
      `bts.transferNativeCoin { value }  (String _to)`
  - WrappedNativeCoin (BNB):
      Get Token Address: `token = bts.coinAddress(String _coinName)`
      Approve: `token.approve(<bts_address>, <approval_amount>)`
      Transfer: `bts.transfer(String _coinName, BigInteger _value, String _to)`
  - IRC2 Tokens:
      Get Token Address: `token = bts.coinAddress(String _coinName)`
      Transfer To BTS: `token.transfer(<bts_address>, <amount>)`
      Transfer: `bts.transfer(String _coinName, BigInteger _value, String _to)`
  - Pre-existing Tokens (IRC2 or otherwise):
    Same as WrappedNativeCoin if the token has approve method, otherwise as IRC2.
    
2. BSC -> ICON:
  - NativeCoin (BNB) :
      `BTSCore.transferNativeCoin { value }  (string calldata _to) external payable`
  - WrappedNativeCoin (ICX):
      Get Token Address: `token = BTSCore.coinId(string calldata _coinName)`
      Approve: `token.approve(<bts_core_address>, <approval_amount>)`
      Transfer: `BTSCore.transfer(string calldata _coinName, uint256 _value, string calldata _to)`
  - Pre-existing Tokens (ERC20):
    Same as WrappedNativeCoin.

**Comment by duyphanz** [comment]
I'd like to know if we have any corresponding changes to the ABI file and the `_to` format also?

e.g:
```
 btp://0x63564c40.hmny/hxeffc184905bfff5db8879914690ba6e5cab2f224
``` 

How about the `balanceOf` and `getBalanceOf` method?

**Reply by bbist** [reply]
Yeah. There have been changes on the ABI, not on the transfer methods but on other helper methods such as `balanceOf` to make it consistent between contracts.

Let me fetch the ABI and forward it to you. There will be some more change on the balanceOf method, but you can start with other methods that haven't changed.

**Reply by duyphanz** [reply]
Hi, @bbist please keep me updated about balanceOf method working schedule as well. Thank you!

**Reply by bbist** [reply]
@izyak, @manishbista28,
Is `balanceOf` method working properly? If so, please share the meaning of the different fields of Balance here.

**Comment by MuhammedIrfan** [comment]
@bbist @CyrusVorwald-ICON I could forsee a chance of a vulnerability, need your thoughts on this.

This vulnerability can be exploited only if admin access is compromised in BTS of any connected chain. 

Consider a scenario where a user creates a HRC20 with Ticker BNB, and managed to get it   registered. if user transfers this token to BTS of BSC wouldn’t that let user drain the BNB.

**Reply by bbist** [reply]
@MuhammedIrfan, I think it would make sense for us to continue this discussion in https://github.com/icon-project/icon-bridge/issues/18.

**Comment by sdpisreddevil** [comment]
ABI for BTS Contracts 

BTS Java
[BTS_JAVA.txt](https://github.com/icon-project/icon-bridge/files/9164216/BTS_JAVA.txt)


BTSPeriphery.sol
[BTSPeriphery-sol.txt](https://github.com/icon-project/icon-bridge/files/9164219/BTSPeriphery-sol.txt)

**Reply by bbist** [reply]
@sdpisreddevil,
Sorry, my mistake. The required ABI was of BTSCore.

**Reply by sdpisreddevil** [reply]
@duyphanz Please check updated ABI for BTS. 

[btpCORE-sol.txt](https://github.com/icon-project/icon-bridge/files/9192475/btpCORE-sol.txt)
[BTS_JAVA.txt](https://github.com/icon-project/icon-bridge/files/9192476/BTS_JAVA.txt)

**Comment by sdpisreddevil** [comment]
[btsCoreAJava.txt](https://github.com/icon-project/icon-bridge/files/9349714/btsCoreAJava.txt)
[btsCoreSol.txt](https://github.com/icon-project/icon-bridge/files/9349715/btsCoreSol.txt)

@duyphanz
